onnxruntime/test/providers/cpu/tensor/gather_elements_op_test.cc

```diff
@@ -75,9 +75,10 @@ void RunTypedTest() {
                      {1, 1,
                       4, 4});
   // skip nuphar, which will not throw error message but will ensure no out-of-bound access
+  // skip cuda as the cuda kernel won't throw the error message
   test5.Run(OpTester::ExpectResult::kExpectFailure,
```

Comment:
Why not? Should/could it?

Reply:
Personally, I am trying to get the answer to this question too.  

Other frameworks like Tf have this kind of operation wherein the CPU op returns an error but in some GPU ops the value if out of bounds is clamped to the nearest valid value and in some ops a 0 is just placed in the output array like here - https://www.tensorflow.org/api_docs/python/tf/gather. So basically the CPU ops throw whereas the GPU ops "succeed" in operation.

We need to be consistent in dealing with these across all GPU ops and we should probably discuss and agree upon a common solution to deal with these.

FWIW - I just followed the precedent set by Scatter Elements - 

https://github.com/microsoft/onnxruntime/blob/d5c36bfff285eba4a0ef7c6b7dabb5b3bf6c5c2e/onnxruntime/core/providers/cuda/tensor/scatter_elements_impl.cu#L31

Of course, we could iterate over the contents of indices once and throw if an invalid value is found but this is a non-starter in the GPU kernel as even valid ops now have to pay the additional "validation cost".

Reply:
It's because CUDA kernel is async call and it will launch hundreds/thousands of CUDA threads.

---
In reply to: [341958959](https://github.com/microsoft/onnxruntime/pull/2310#discussion_r341958959) [](ancestors = 341958959)